lib/client.js

```diff
@@ -1,55 +1,131 @@
-    // convert arguments to real array
-    const args = Array.prototype.slice.call(arguments, 0);
-    const cb = args.pop();
-
-    // Add "time" between command and rest of args
-    const time = new Date().getTime() / 1000.0;
-    args.splice(1, 0, time);
-    this._lua.call(args, cb);
-  }
+// Return a new disposer
+Client.disposer = function(config) {
+  return new bluebird.Promise((resolve, reject) => {
+    resolve(new Client(config));
+  }).disposer((client, promise) => {
+    client.quit();
+  });
+}
+
+// Use a client only for the duration of handler
+Client.using = function(config, handler) {
+  return bluebird.Promise.using(Client.disposer(config), handler);
+}
+
+// Close this client
+Client.prototype.quit = function() {
+  logger.debug('Closing client.');
+  this.redis.quit();
+}
 
-  /**
-   * Get a Queue object for the given queue name
-   */
-  queue(name) {
-    // Note: As with the Python ForkingWorker, all workers under one process
-    // have the same workerName. Dan says: "They do occasionally step on each others'
-    // toes (one process loses its lock on a job when the other pops it off, then both
-    // are notified that the lock is lost and second one drops it). The upside to this
-    // design decision is that we have fewer unique worker names and they're not
-    // constantly changing. In practice, this happens infrequently and negligibly
-    // impacts the delivery of useful work." See https://github.com/seomoz/qless-js/pull/4
-    return new queue.Queue(name, this, this.workerName);
+// Invoke a qless command
+Client.prototype.call = function(command, ...args) {
+  let now = (new Date).getTime() / 1000.0;
+  let self = this;
+  return this.load()
+    .then((sha) => {
+      logger.debug('Calling %s', command, {
+        now: now,
+        args: args,
+      });
+      return self.redis.evalshaAsync([sha, 0, command, now, ...args]);
+    })
+    .then((response) => {
+      logger.debug('Response: %s', response);
+      return response;
+    });
+}
+
+// Load the Lua scripts
+Client.prototype.load = function() {
+  let self = this;
+
+  return new bluebird.Promise((resolve, reject) => {
+    // Short-circuit if we've already saved the SHA
+    if (self.sha !== null) {
+      return resolve(self.sha);
+    }
+
+    // Read and memoize the SHA
+    fs.readFileAsync(this.path)
+      .then(function(contents) {
+        return self.redis.scriptAsync('load', contents);
+      })
+      .then(function(sha) {
+        logger.debug('Loaded SHA %s', sha);
+        self.sha = sha;
+        return resolve(self.sha);
+      })
+      .catch(reject)
+  });
+}
+
+// Get a specific job by id
+Client.prototype.job = function(jid) {
+  let self = this;
+  return this.call('get', jid)
+    .then((job) => {
+      if (job) {
+        // It's a normal job
+        return new Job(self, JSON.parse(job));
+      } else {
+        // Try to get a recurring job
+        return self.call('recur.get', jid)
+          .then((job) => {
```

Comment:
Yeah, I agree. The linting complained about the reuse of the name `job`, so I ended up going with `recurringJob`, which made the single line a bit unwieldy.